modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js

```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import ClayButton from '@clayui/button';
+import {Align, ClayDropDownWithItems} from '@clayui/drop-down';
+import ClayIcon from '@clayui/icon';
+import ClaySticker from '@clayui/sticker';
+import {fetch} from 'frontend-js-web';
+import PropTypes from 'prop-types';
+import React, {createRef, useState} from 'react';
+
+const preloadPromise = createRef();
+
+function PersonalMenu({itemsURL}) {
+	const [active, setActive] = useState(false);
+	const [items, setItems] = useState([]);
+
+	function preloadItems() {
+		if (!preloadPromise.current) {
+			preloadPromise.current = fetch(itemsURL)
+				.then(response => response.json())
+				.then(items => setItems(items));
```

Comment:
It's not good practice to call a `set` method asynchronously if there's a chance that the component got unmounted before the callback fires (and even if you think it never will get unmounted, it is a brittle assumption). So you can guard this with a `isMounted` check:

```javascript
import {useIsMounted} from 'frontend-js-react-web`;
```

Then:

```javascript
const isMounted = useIsMounted();

// and in the then ....
then(items => {
  if (isMounted()) {
    setItems(items);
  }
});
```

In layout-content-page-editor-web I made [a `useStateSafe` hook](https://github.com/brianchandotcom/liferay-portal/blob/cd6ad8bc54f4a437abfc6fedbcfb338a877e6f0d/modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/core/hooks/useStateSafe.js) that can be used as a drop-in replacement for `useState` that bakes this check in. I think we just need to make sure the naming is good (I don't know that it is) and we can move it into frontend-js-react-web.

Reply:
Thanks! I've heard this so many times and yet forgot to apply it! 😢